pulse-rs/src/context.rs

```diff
@@ -0,0 +1,398 @@
+            let ctx = context::from_raw_ptr(c);
+            let result = ::std::mem::transmute::<_, &F>(&())(&ctx, info, eol, userdata);
+            ::std::mem::forget(ctx);
+            result
+        }
+
+        let o = unsafe {
+            ffi::pa_context_get_source_info_list(self.raw_mut(), Some(wrapped::<CB>), userdata)
+        };
+        if o.is_null() {
+            return Err(ErrorCode::from_error_code(self.errno()));
+        }
+        Ok(unsafe { ::operation::from_raw_ptr(o) })
+    }
+
+    pub fn set_sink_input_volume<CB>(&self,
+                                     idx: u32,
+                                     volume: &CVolume,
+                                     _: CB,
+                                     userdata: *mut c_void)
+                                     -> Result<Operation>
+        where CB: Fn(&Context, i32, *mut c_void)
+    {
+        debug_assert_eq!(::std::mem::size_of::<CB>(), 0);
+
+        unsafe extern "C" fn wrapped<F>(c: *mut ffi::pa_context,
+                                        success: c_int,
+                                        userdata: *mut c_void)
+            where F: Fn(&Context, i32, *mut c_void)
+        {
+            let ctx = context::from_raw_ptr(c);
+            let result = ::std::mem::transmute::<_, &F>(&())(&ctx, success, userdata);
+            std::mem::forget(ctx);
+            result
+        }
+
+        let o = unsafe {
+            ffi::pa_context_set_sink_input_volume(self.raw_mut(),
+                                                  idx,
+                                                  volume,
+                                                  Some(wrapped::<CB>),
+                                                  userdata)
+        };
+        if o.is_null() {
+            return Err(ErrorCode::from_error_code(self.errno()));
+        }
+        Ok(unsafe { operation::from_raw_ptr(o) })
+    }
+
+    pub fn subscribe<CB>(&self,
+                         m: SubscriptionMask,
+                         _: CB,
+                         userdata: *mut c_void)
+                         -> Result<Operation>
+        where CB: Fn(&Context, i32, *mut c_void)
+    {
+        debug_assert_eq!(::std::mem::size_of::<CB>(), 0);
+
+        unsafe extern "C" fn wrapped<F>(c: *mut ffi::pa_context,
+                                        success: c_int,
+                                        userdata: *mut c_void)
+            where F: Fn(&Context, i32, *mut c_void)
+        {
+            let ctx = context::from_raw_ptr(c);
+            let result = ::std::mem::transmute::<_, &F>(&())(&ctx, success, userdata);
+            std::mem::forget(ctx);
+            result
+        }
+
+        let o = unsafe {
+            ffi::pa_context_subscribe(self.raw_mut(), m.into(), Some(wrapped::<CB>), userdata)
+        };
+        if o.is_null() {
+            return Err(ErrorCode::from_error_code(self.errno()));
+        }
+        Ok(unsafe { operation::from_raw_ptr(o) })
+    }
```

Comment:
I didn't know how to access self from macros. I read up and I just have to pass self in as a param.